tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc

```diff
@@ -500,6 +501,80 @@ DiagnosedSilenceableFailure ApplyPatternsOp::applyToOne(
   return DiagnosedSilenceableFailure(success());
 }
 
+//===---------------------------------------------------------------------===//
+// FoldProducerExtractSliceOp
+//===---------------------------------------------------------------------===//
+
+namespace {
+
+LogicalResult tryMergeExtractSliceOp(OpBuilder& b,
+                                     tensor::ExtractSliceOp& sliceOp,
+                                     tensor::ExtractSliceOp producerOp) {
+  if (!strideAllOnes(sliceOp) || !strideAllOnes(producerOp)) return failure();
+  Location loc = sliceOp->getLoc();
+  SmallVector<OpFoldResult> newOffsets;
+  for (const auto& z :
+       llvm::zip(sliceOp.getMixedOffsets(), producerOp.getMixedOffsets())) {
+    Value lhsValue = std::get<0>(z).dyn_cast<Value>();
+    Value rhsValue = std::get<1>(z).dyn_cast<Value>();
+    if (!lhsValue) {
+      lhsValue = b.create<arith::ConstantIndexOp>(
+          loc,
+          std::get<0>(z).dyn_cast<Attribute>().cast<IntegerAttr>().getInt());
+    }
+    if (!rhsValue) {
+      rhsValue = b.create<arith::ConstantIndexOp>(
+          loc,
+          std::get<1>(z).dyn_cast<Attribute>().cast<IntegerAttr>().getInt());
+    }
+    newOffsets.push_back(
+        b.create<arith::AddIOp>(loc, lhsValue, rhsValue).getResult());
+  }
+  sliceOp = b.create<tensor::ExtractSliceOp>(
+      loc, producerOp.getSource(), newOffsets, sliceOp.getMixedSizes(),
+      sliceOp.getMixedStrides());
+  return success();
+}
+
+}  // namespace
+
+void FoldProducerExtractSliceOp::build(OpBuilder& builder,
+                                       OperationState& result, Value target,
+                                       int64_t maxRepeatNum) {
+  MLIRContext* ctx = builder.getContext();
+  result.addOperands(target);
+  result.addAttribute(
+      FoldProducerExtractSliceOp::getMaxRepeatNumAttrName(result.name),
+      builder.getIntegerAttr(builder.getIntegerType(64), maxRepeatNum));
+  result.addTypes({pdl::OperationType::get(ctx)});
+}
+
+DiagnosedSilenceableFailure FoldProducerExtractSliceOp::applyToOne(
```

Comment:
newbie question: where is this method called? Is this a override method?

Reply:
It's defined by `TransformEachOpTrait` which is one of the traits `FoldProducerExtractSliceOp` has.